lib/bundler/definition.rb

```diff
@@ -54,7 +54,8 @@ def self.build(gemfile, lockfile, unlock)
     def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, optional_groups = [])
       @unlocking = unlock == true || !unlock.empty?
 
-      @dependencies    = dependencies
+      @dependencies = dependencies
+      @dependencies_by_name = dependencies.group_by(&:name)
```

Comment:
I'm a little concerned if this would introduce a boot time regression. I can take a look in a bit and test it out